### PR TITLE
Add Get/SetBlockIoTune to go-libvirt API.

### PR DESCRIPTION
**Pull request by trapgate**
This adds two libvirt entry points to the go-libvirt API:
virDomainSetBlockIoTune and virDomainGetBlockIoTune. These can be used
to control block device throttling for a VM.

### CODE REVIEW

File: libvirt.go [hunk 1]
```diff
@@ -94,6 +95,75 @@ type qemuError struct {
 // DomainXMLFlags specifies options for dumping a domain's XML.
 type DomainXMLFlags uint32
 
+// Consts used for flags
+
+// virDomainModificationImpact and virTypedParameterFlags
```
Comment (benlemasurier) [inline libvirt.go]: can this comment be more descriptive than a reference to libvirt's source?

File: libvirt.go [hunk 2]
```diff
@@ -94,6 +95,75 @@ type qemuError struct {
...
+const (
```
Comment (benlemasurier) [inline libvirt.go]: let's add a new type, something like `DomainAffectFlags uint32`, then

```go
FlagDomainAffectCurrent DomainAffectFlags = 0
...
```

File: libvirt.go [hunk 3]
```diff
@@ -94,6 +95,75 @@ type qemuError struct {
...
+	FlagDomainAffectCurrent = 0
+	FlagDomainAffectLive    = 1 << (iota - 1)
+	FlagDomainAffectConfig
+	FlagTypedParamStringOkay
+)
+
+// Consts relating to TypedParams:
+const (
+	_ = iota
```
Comment (benlemasurier) [inline libvirt.go]: this can go away, using `iota + 1` on the next line

File: libvirt.go [hunk 4]
```diff
@@ -94,6 +95,75 @@ type qemuError struct {
...
+	TypedParamInt
+	// TypedParamUInt is a C unsigned int.
+	TypedParamUInt
+	// TypedParamLLong is a C long long int.
+	TypedParamLLong
+	// TypedParamULLong is a C unsigned long long int.
+	TypedParamULLong
+	// TypedParamDouble is a C double.
+	TypedParamDouble
+	// TypedParamBoolean is a c char.
```
Comment (benlemasurier) [inline libvirt.go]: `C char`

File: libvirt.go [hunk 5]
```diff
@@ -94,6 +95,75 @@ type qemuError struct {
...
+	TypedParamBoolean
+	// TypedParamString is a c char*.
```
Comment (benlemasurier) [inline libvirt.go]: `C char*`

File: libvirt.go [hunk 6]
```diff
@@ -1199,6 +1269,128 @@ func (l *Libvirt) Reset(dom string) error {
...
+	if err != nil {
+		return err
+	}
+
+	// https://libvirt.org/html/libvirt-libvirt-domain.html#virDomainSetBlockIoTune
+	payload := struct {
+		Domain Domain
+		Disk   string
+		Params []TypedParam
+		Flags  uint32
```
Comment (benlemasurier) [inline libvirt.go]: `Flags DomainAffectFlags` or whatever you might have decided to call it :)

File: libvirttest/libvirt.go [hunk 7]
```diff
@@ -399,6 +400,80 @@ var testRebootReply = []byte{
...
+// standard header to the beginning. The length parameter has to be correct!
+var testGetBlockIoTuneReply = []byte{
+	0x00, 0x00, 0x02, 0xe0, // length
+	0x20, 0x00, 0x80, 0x86, // program
+	0x00, 0x00, 0x00, 0x01, // version
+	0x00, 0x00, 0x00, 0xfd, // procedure
+	0x00, 0x00, 0x00, 0x00, // type
+	0x00, 0x00, 0x00, 0x00, // serial
+	0x00, 0x00, 0x00, 0x00, // status
+	0x0, 0x0, 0x0, 0x13, 0x0, 0x0, 0x0, 0xf,
```
Comment (benlemasurier) [inline libvirttest/libvirt.go]: can we roughly comment on the data contained below?

File: libvirt.go [hunk 8]
```diff
@@ -1199,6 +1269,128 @@ func (l *Libvirt) Reset(dom string) error {
...
+	Name  string
+	Value uint64
+}
+
+// SetBlockIoTune changes the per-device block I/O tunables within a guest.
+// Parameters are the name of the VM, the name of the disk device to which the
+// limits should be applied, and 1 or more BlockLimit structs containing the
+// actual limits.
+//
+// Possible limits which can be applied here include:
```
Comment (benlemasurier) [inline libvirt.go]: Let's make these constants

File: libvirt.go (hunk 1 above)
Comment (trapgate) [reply libvirt.go]: Added more details, and comments for each of the consts too.

File: libvirt.go (hunk 2 above)
Comment (trapgate) [reply libvirt.go]: Done.

File: libvirt.go (hunk 3 above)
Comment (trapgate) [reply libvirt.go]: Cool, I like that better.

File: libvirt.go (hunk 4 above)
Comment (trapgate) [reply libvirt.go]: Fixed

File: libvirt.go (hunk 5 above)
Comment (trapgate) [reply libvirt.go]: Fixed

File: libvirttest/libvirt.go (hunk 7 above)
Comment (trapgate) [reply libvirttest/libvirt.go]: I can, yes. In general though, it's more valuable to maintainers to tell them how to get these bytes than how to change them - we don't really want devs to construct these test blocks themselves, as that invalidates the test. I'm guessing that tcpdump was used to capture the first few, but some of them have the wrong values for procedure, so you can tell they weren't captured.

File: libvirt.go [hunk 9]
```diff
@@ -1199,6 +1280,128 @@ func (l *Libvirt) Reset(dom string) error {
...
+	Value uint64
+}
+
+// SetBlockIOTune changes the per-device block I/O tunables within a guest.
+// Parameters are the name of the VM, the name of the disk device to which the
+// limits should be applied, and 1 or more BlockLimit structs containing the
+// actual limits.
+//
+// Possible limits which can be applied here include:
+// - total_bytes_sec
```
Comment (benlemasurier) [inline libvirt.go]: missed a previous comment here? Can we make these constants?

File: internal/constants/constants.go [hunk 10]
```diff
@@ -53,6 +53,8 @@ const (
 	ProcDomainUndefineFlags        = 231
 	ProcDomainDestroyFlags         = 234
 	ProcDomainReset                = 245
+	ProcDomainSetBlockIoTune       = 252
```
Comment (benlemasurier) [inline internal/constants/constants.go]: `ProcDomainSetBlockIOTune`
`ProcDomainGetBlockIOTune`

File: libvirt_test.go [hunk 11]
```diff
@@ -437,3 +437,27 @@ func TestReset(t *testing.T) {
 		t.Fatalf("unexpected reset error: %v", err)
 	}
 }
+
+func TestSetBlockIoTune(t *testing.T) {
```
Comment (benlemasurier) [inline libvirt_test.go]: `IOTune`

File: libvirt_test.go [hunk 12]
```diff
@@ -437,3 +437,27 @@ func TestReset(t *testing.T) {
...
+	conn := libvirttest.New()
+	l := New(conn)
+
+	if err := l.SetBlockIOTune("test", "vda", BlockLimit{"write_bytes_sec", 5000000}); err != nil {
+		t.Fatalf("unexpected SetBlockIoTune error: %v", err)
+	}
+}
+
+func TestGetBlockIoTune(t *testing.T) {
```
Comment (benlemasurier) [inline libvirt_test.go]: `IOTune`

File: libvirttest/libvirt.go [hunk 13]
```diff
@@ -399,6 +400,128 @@ var testRebootReply = []byte{
 	0x00, 0x00, 0x00, 0x00, // status
 }
 
+var testSetBlockIoTuneReply = []byte{
```
Comment (benlemasurier) [inline libvirttest/libvirt.go]: `IOTune`

File: libvirttest/libvirt.go [hunk 14]
```diff
@@ -399,6 +400,128 @@ var testRebootReply = []byte{
...
+	0x00, 0x00, 0x00, 0xfc, // procedure
+	0x00, 0x00, 0x00, 0x00, // type
+	0x00, 0x00, 0x00, 0x00, // serial
+	0x00, 0x00, 0x00, 0x00, // status
+}
+
+// This result block was obtained by calling `fmt.Printf("%#v", r.Payload)` on
+// the result returned by an actual call to GetBlockIoTune, and then adding the
+// standard header to the beginning. The length parameter has to be correct!
+var testGetBlockIoTuneReply = []byte{
```
Comment (benlemasurier) [inline libvirttest/libvirt.go]: `IOTune`

File: libvirttest/libvirt.go [hunk 15]
```diff
@@ -486,6 +609,10 @@ func (m *MockLibvirt) handleRemote(procedure uint32, conn net.Conn) {
 		conn.Write(m.reply(testCreateWithFlags))
 	case constants.ProcDomainShutdownFlags:
 		conn.Write(m.reply(testShutdownReply))
+	case constants.ProcDomainSetBlockIoTune:
```
Comment (benlemasurier) [inline libvirttest/libvirt.go]: `IOTune`

**Comment by trapgate** [comment]
All Io's are now embiggened. Sorry about that.

File: libvirt.go (hunk 9 above)
Comment (trapgate) [reply libvirt.go]: Wildo. I'm looking for the source from which they might eventually be generated.

Comment (trapgate) [reply libvirt.go]: Done.

**Pull request merged**